self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryContracts.java

```diff
@@ -22,43 +22,392 @@
+            }
+            projectInternal = new ProjectInternal(projectFromStorage, this);
+            projects.add(projectInternal);
+        }
+
+        return projectInternal;
     }
+
+    /**
+     * Returns a contributor by id from the contributors set.
+     * If the contributor is not found, a contributor will
+     * be created and added to set.
+     *
+     * @param cbId Contributor's id
+     * @return Internal representation of a Contributor
+     */
+    private ContributorInternal findOrCreateContributor(final int cbId) {
+        ContributorInternal contributor = contributors.stream()
+            .filter(c -> c.contributorId() == cbId)
+            .findFirst()
+            .orElse(null);
+        if (contributor == null) {
+            contributor = new ContributorInternal(cbId, this);
+            contributors.add(contributor);
+        }
+        return contributor;
+    }
+
+
+    /**
+     * In memory implementation of a {@link Project}.
+     * Wraps a real implementation of a Project.
+     * <br>
+     * Aids {@link InMemoryContracts}
+     * <br>
+     * Has capability to link a contract.
+     * See: {@link ProjectInternal#addContract(Contract)}.
+     */
+    private static final class ProjectInternal implements Project {
+
+        /**
+         * Real implementation of a project.
+         */
+        private final Project delegate;
+        /**
+         * Global contracts. Access to {@link InMemoryContracts }
+         * Grants access to any arbitrary project's contracts using
+         * {@link Contracts#ofProject(int) from any point of the object graph.
+         */
+        private final Contracts rootContracts;
+
+        /**
+         * Contracts associated to this project.
+         */
+        private final Set<Contract> ownedContracts = new HashSet<>();
+
+        /**
+         * Constructor.
+         *
+         * @param delegate Real implementation of a Project
+         * @param rootContracts Global contracts.
+         *                      Access to {@link InMemoryContracts }
+         */
+        ProjectInternal(final Project delegate,
+                        final Contracts rootContracts) {
+            this.delegate = delegate;
+            this.rootContracts = rootContracts;
+        }
+
+        /**
+         * Adds a contract to ownedContracts.
+         *
+         * @param contract Contract to be added.
+         */
+        void addContract(final Contract contract) {
+            ownedContracts.add(contract);
+        }
+
+        @Override
+        public int projectId() {
+            return delegate.projectId();
+        }
+
+        @Override
+        public ProjectManager projectManager() {
+            return delegate.projectManager();
+        }
+
+        @Override
+        public Repo repo() {
+            return delegate.repo();
+        }
+
+        @Override
+        public Contracts contracts() {
+            return new Contracts() {
```

Comment:
@criske I think you can simply return ``rootContracts.ofProject(...)``

Reply:
We need the "owned" contracts when iterate this project contracts, that's why I've used this proxy. 
rootContracts.ofProject is "global" search for any valid project id

Reply:
Also if we  just call rootContracts.ofProject(this.projectId()) instead, we will get stack overflow